src/components/utils/test/timer_test.cc

```diff
@@ -43,214 +43,276 @@ namespace components {
 namespace timer_test {
 namespace {
 
-sync_primitives::Lock test_lock;
-sync_primitives::ConditionalVariable lock_monitor;
+sync_primitives::Lock shot_lock;
+sync_primitives::ConditionalVariable shot_condition;
+
+const std::string kTimerName = "TestTimer";
+const bool kSingleShot = true;
 const uint32_t kDefaultTimeout = 30u;
-const std::string kTimerName = "test_timer";
+const uint32_t kDefaultTimeoutRestart = 45u;
 
 class TestTask : public timer::TimerTask {
  public:
-  TestTask() : calls_count_(0u) {}
+  TestTask() : timer_(NULL), calls_count_(0u) {}
+
+  void set_timer(timer::Timer* timer) {
+    timer_ = timer;
+  }
+
+  virtual void PerformTimer() const {}
 
   void run() const OVERRIDE {
-    sync_primitives::AutoLock auto_lock(test_lock);
+    sync_primitives::AutoLock auto_lock(shot_lock);
     ++calls_count_;
-    lock_monitor.NotifyOne();
+    shot_condition.NotifyOne();
+    PerformTimer();
   }
 
-  uint GetCallsCount() const {
+  size_t calls_count() const {
     return calls_count_;
   }
 
- private:
-  mutable uint calls_count_;
+ protected:
+  mutable timer::Timer* timer_;
```

Comment:
@malirod for special case with timer `Start`/`Stop` from callback. Callback should be `const`. As for me it's acceptable for testing purposes.